File: R/PANPRS_gsfPEN.R
#' Run the gsfPEN algorithm for multiple traits, with functional annotations.
#' @param summary_z A matrix of summary statistics for each SNP and trait.
#' @param n_vec A vector of sample sizes for each of the Q traits corresponding to the Q columns of summary_z.
#' @param plinkLD A matrix of LD values for each pair of SNPs.
#' @param func_index A matrix of functional annotations for each SNP and trait. For the element at i-th row, j-th column, the entry 0 means SNP i without j-th functional annotation; entry 1 means otherwise.
#' @param n_iter The number of iterations to run the algorithm.
#' @param upper_val The upper bound for the tuning parameter.
#' @param breaking The number of iterations to run before checking for convergence.
#' @param z_scale The scaling factor for the summary statistics.
#' @param tuning_matrix A matrix of tuning parameters.
#' @param p_threshold A vector of p-values to use for the tuning parameters.
#' @param p_threshold_params A vector of parameters to use for the p-value tuning parameters.
#' @param tau_factor A vector of factors to multiply the median value by to get the tuning parameters.
#' @param sub_tuning The number of tuning parameters to use for the second iteration.
#' @param lim_lambda The range of tuning parameters to use for the first iteration.
#' @param len_lambda The number of tuning parameters to use for the second iteration.
#' @param lambda_vec A vector of tuning parameters to use for the first iteration.
#' @param lambda_vec_limit_len The number of tuning parameters to use for the first iteration.
#' @param df_max The maximum degrees of freedom for the model.
#' @param sparse_beta Whether to use the sparse version of the algorithm.
#' @param debug_output Whether to output the tuning combinations that did not converge.
#' @param verbose Whether to output information through the evaluation of the algorithm.
#' @return A named list containing the following elements:
#' beta_matrix: A matrix of the estimated coefficients for each SNP and trait.
#' num_iter_vec: A vector of the number of iterations for each tuning combination.
#' all_tuning_matrix: A matrix of the tuning parameters used for each tuning combination.
#' @importFrom stats median qnorm quantile
#' @examples
#' # Load the library and data
#' library(PANPRSnext)
#' data("summaryZ")
#' data("Nvec")
#' data("plinkLD")
#' data("funcIndex")
#'
#' # Take random subset of the data
#' subset <- sample(nrow(summaryZ), 100)
#' subset_summary_z <- summaryZ[subset, ]
#' subset_func_index <- funcIndex[subset, ]
#'
#' # Run gsfPEN
#' output <- gsfPEN_R(
#'   summary_z = subset_summary_z,
#'   n_vec = Nvec,
#'   plinkLD = plinkLD,
#'   func_index = subset_func_index
#' )
#' @export
gsfPEN_R <- function(
    summary_z,
    n_vec,
    plinkLD,
    func_index,
    n_iter = 1000,
    upper_val = NULL,
    breaking = 1,
    z_scale = 1,
    tuning_matrix = NULL,
    p_threshold = NULL,
    p_threshold_params = c(0.5, 10^-4, 4),
    tau_factor = c(1 / 25, 1, 3),
    sub_tuning = 4,
    lim_lambda = c(0.5, 0.9),
    len_lambda = 4,
    lambda_vec = NULL,
    lambda_vec_limit_len = c(1.5, 3),
    df_max = NULL,
    sparse_beta = FALSE,
    debug_output = FALSE,
    verbose = FALSE) {
  if (z_scale != 1) {
    stop("Tuning values set-up for multiple traits analysis requires z_scale = 1") # nolint: object_usage_linter.
  }

  num_func <- ncol(func_index)

  P <- nrow(summary_z)
  Q <- length(n_vec)

  summary_betas <- matrix(0, nrow = P, ncol = Q)
  SD_vec <- matrix(0, nrow = P, ncol = Q)

  for (i in 1:Q) {
    summary_betas[, i] <- summary_z[, i] / sqrt(n_vec[i])
    SD_vec[, i] <- 1 / sqrt(n_vec[i])
  }

  rownames(summary_betas) <- rownames(summary_z)

  if (is.null(df_max)) {
    df_max <- ceiling(0.7 * P)
  }

  if (is.null(p_threshold)) {
    p_threshold <- seq(p_threshold_params[1], p_threshold_params[2], length.out = p_threshold_params[3])
  }

  if (any(c(is.null(tuning_matrix), is.null(lambda_vec)))) {
    median_val <- median(apply(abs(summary_betas), 1, sum), na.rm = TRUE)
    tau_vec <- sort(median_val * tau_factor)
    lim_lambda <- quantile(abs(summary_z[, 1]), lim_lambda)

    output <- Tuning_setup_group_func(
      lambda_vec,
      lambda_vec_limit_len,
      #   p_threshold,
      num_func,
      tau_vec,
      sub_tuning,
      lim_lambda,
      len_lambda,
      # len_lim_lambda,
      median_val,
      equal_lambda = TRUE,
      Q
    )

    func_lambda <- output$func_lambda
    lambda_vec <- output$lambda_vec
    tuning_matrix <- output$tuning_matrix

    rm(output)
  } else {
    func_lambda <- permutations( # nolint: object_usage_linter.
      length(lambda_vec),
      num_func,
      repeats.allowed = TRUE
    ) - 1
  }

  beta_index <- c(seq_len(nrow(summary_betas))) - 1
  SNP_names <- rownames(summary_betas)

  # Takes only the SNPs that are present in both the PLINK data set and the GWAs
  ld_J <- PlinkLD_transform(plinkLD, SNP_names)

  J_index_matrix <- matrix(nrow = nrow(ld_J), ncol = 2)

  mat1 <- match(ld_J[, 1], SNP_names)
  mat2 <- match(ld_J[, 2], SNP_names)

  J_index_matrix[, 1] <- beta_index[mat1]
  J_index_matrix[, 2] <- beta_index[mat2]

  ld_J[, 1] <- J_index_matrix[, 1]
  ld_J[, 2] <- J_index_matrix[, 2]

  od <- order(J_index_matrix[, 1], J_index_matrix[, 2], decreasing = FALSE)
  ld_J <- ld_J[od, ]


  wind <- which(!beta_index %in% ld_J[, 1])
  num_indices <- length(wind)

  index_J <- -1
  if (num_indices > 0) {
    index_J <- beta_index[wind]
  }

  counts <- table(ld_J[, 1])
  num_SNP <- length(counts)

  index_S <- c(0, cumsum(counts)[-num_SNP])
  index_E <- cumsum(counts) - 1

  index_matrix <- matrix(nrow = num_SNP, ncol = 3)
  index_matrix[, 1] <- as.numeric(names(counts))

  index_matrix[, 2] <- index_S
  index_matrix[, 3] <- index_E

  ld_vec <- ld_J[, 3]
  ld_J <- ld_J[, 2]


  if (is.null(upper_val)) {
    upper_val <- ceiling(max(abs(summary_betas), na.rm = TRUE) * 50)
  }

  if (nrow(func_index) != nrow(summary_betas)) {
    stop("nrow of summary_betas and row of func_index do not match")
  }


  z_matrix <- as.matrix(1 - func_index)

  sum_func_index <- apply(z_matrix, 1, sum)
  Ifunc_SNP <- rep(0, P)
  Ifunc_SNP[which(sum_func_index != 0)] <- 1

  lambda0_vec <- abs(-qnorm(p_threshold / 2))

  nrow_index_matrix <- nrow(index_matrix)
  ncol_index_matrix <- ncol(index_matrix)

  nrow_z_matrix <- nrow(z_matrix)
  ncol_z_matrix <- ncol(z_matrix)

  nrow_func_lambda <- nrow(func_lambda)
  ncol_func_lambda <- ncol(func_lambda)

  nrow_tuning_matrix <- nrow(tuning_matrix)
  ncol_tuning_matrix <- ncol(tuning_matrix)

  nrow_all_tuning_matrix <- nrow_tuning_matrix * length(p_threshold) * nrow_func_lambda
  ncol_all_tuning_matrix <- (num_func + 1) + 2

  nrow_beta_matrix <- nrow_all_tuning_matrix
  ncol_beta_matrix <- P * Q

  dims <- c(
    num_SNP, # 1
    P, # 2
    Q, # 3
    nrow_index_matrix, # 4
    ncol_index_matrix, # 5
    nrow_z_matrix, # 6
    ncol_z_matrix, # 7
    nrow_func_lambda, # 8
    ncol_func_lambda, # 9
    nrow_tuning_matrix, # 10
    ncol_tuning_matrix, # 11
    nrow_all_tuning_matrix, # 12 == num_tuning
    ncol_all_tuning_matrix, # 13
    nrow_beta_matrix, # 14
    ncol_beta_matrix # 15
  )

  params <- c(
    upper_val, # 1
    n_iter, # 2
    breaking, # 3
    z_scale, # 4
    df_max, # 5
    length(p_threshold), # 6
    num_indices # 7
  )

  args <- list(
    summary_betas,
    ld_J,
    index_matrix,
    index_J,
    ld_vec,
    SD_vec,
    tuning_matrix,
    lambda0_vec,
    z_matrix,
    lambda_vec,
    func_lambda,
    Ifunc_SNP,
    dims,
    params
  )

  if (verbose) {
    cat("Number of total tuning combinations =", nrow_all_tuning_matrix, "\n")
  }

  if (sparse_beta) {
    Z <- do.call(
      gsfPEN_sparse_cpp,
      args
    )
  } else {
    Z <- do.call(
      gsfPEN_cpp,
      args
    )
  }

  beta_matrix <- Z$beta_matrix
  colnames(beta_matrix) <- paste0(
    rep(SNP_names, times = Q),
    ".trait",
    rep(c(1:Q), each = P)
  )

  all_tuning_matrix <- Z$all_tuning_matrix
  colnames(all_tuning_matrix) <- c(
    "lambda0",
    paste0("lambdaf", c(1:num_func)),
    "lambda2",
    "tau2"
  )

  num_iter_vec <- Z$num_iter_vec

  # Remove the tuning combinations that did not converge (correspons to -2 in num_iter_vec)
  if (!debug_output) {
    converge_index <- which(num_iter_vec > 0)
    if (verbose) {
      cat("Removing (", length(num_iter_vec) - length(converge_index), ") tuning combinations that did not converge\n")
    }
    num_iter_vec <- num_iter_vec[converge_index]
    beta_matrix <- beta_matrix[converge_index, ]
    all_tuning_matrix <- all_tuning_matrix[converge_index, ]
  }

  output <- Clean_results(
    beta_matrix = beta_matrix,
    num_iter_vec = num_iter_vec,
    all_tuning_matrix = all_tuning_matrix
  )

  return(output)
}

File: src/PANPRS_gsfPEN.cpp
// Prevent bounds checking for increased performance
#define ARMA_NO_DEBUG 1

#include <RcppArmadillo.h>

#include <stdio.h>
#include <stdlib.h>

#include "PANPRS_gsfPEN.h"

Rcpp::List gsfPEN_cpp(
    arma::Mat<double> summary_betas,
    arma::Col<int> ld_J,
    arma::Mat<int> index_matrix,
    arma::Col<int> index_J,
    arma::Col<double> ld_vec,
    arma::Mat<double> SD_vec,
    arma::Mat<double> tuning_matrix,
    arma::Col<double> lambda0_vec,
    arma::Mat<double> z_matrix,
    arma::Col<double> lambda_vec,
    arma::Mat<int> func_lambda,
    arma::Col<int> Ifunc_SNP,
    arma::Col<int> dims,
    arma::Col<double> params)
{
  int num_SNP = dims(0);

  int P = dims(1);
  int Q = dims(2);

  int nrow_func_lambda = dims(7);
  int ncol_func_lambda = dims(8);

  int nrow_tuning_matrix = dims(9);

  int nrow_all_tuning_matrix = dims(11);
  int ncol_all_tuning_matrix = dims(12);

  int nrow_beta_matrix = dims(13);
  int ncol_beta_matrix = dims(14);

  double upper_val = params(0);
  int num_iter = params(1);
  int breaking = params(2);
  int z_scale = params(3);
  int df_max = params(4);
  int leng_p_threshold = params(5);
  int num_indices = params(6);

  double epsilon = 0.0001;

  arma::Col<int> num_iter_vec(nrow_all_tuning_matrix, arma::fill::zeros);
  arma::Col<double> temp_lambda_vec(ncol_func_lambda, arma::fill::zeros);
  arma::Col<double> sum_betas(P, arma::fill::zeros);

  arma::Mat<double> all_tuning_matrix(nrow_all_tuning_matrix, ncol_all_tuning_matrix, arma::fill::zeros);
  arma::Mat<double> beta_matrix(nrow_beta_matrix, ncol_beta_matrix, arma::fill::zeros);
  arma::Mat<double> joint_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<double> temp_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<int> skip(P, Q, arma::fill::zeros);

  GetRNGstate();

  int tuning_index = -1;
  for (int threshold_index = 0; threshold_index < leng_p_threshold; threshold_index++)
  {
    for (int tun_idx_1 = 0; tun_idx_1 < nrow_func_lambda; tun_idx_1++)
    {
      for (int tun_idx_2 = 0; tun_idx_2 < nrow_tuning_matrix; tun_idx_2++)
      {
        tuning_index++;

        for (int func_index = 0; func_index < ncol_func_lambda; func_index++)
        {
          temp_lambda_vec(func_index) = lambda_vec(func_lambda(tun_idx_1, func_index));
          all_tuning_matrix(tuning_index, func_index + 1) = temp_lambda_vec(func_index);
        }

        double lambda0 = lambda0_vec(threshold_index);
        all_tuning_matrix(tuning_index, 0) = lambda0;

        double lambda2 = tuning_matrix(tun_idx_2, 2);
        all_tuning_matrix(tuning_index, ncol_all_tuning_matrix - 2) = lambda2;

        double tau2 = tuning_matrix(tun_idx_2, 3);
        all_tuning_matrix(tuning_index, ncol_all_tuning_matrix - 1) = tau2;

        joint_b_matrix.zeros();
        temp_b_matrix.zeros();
        skip.zeros();
        sum_betas.zeros();

        bool converges = true;
        for (int n = 1; n <= num_iter; n++)
        {
          if (num_indices != 0)
          {
            for (int i = 0; i < num_indices; i++)
            {
              int j = index_J(i);

              double lambda1 = lambda0 + (Ifunc_SNP(j) == 1) * arma::as_scalar(z_matrix.row(j) * temp_lambda_vec);

              for (int q = 0; q < Q; q++)
              {
                double bj_bar = summary_betas(j, q);
                if (bj_bar != 0.0)
                {
                  double threshold = ((z_scale == 1) * SD_vec(j, q) + (z_scale == 0)) *
                                     (lambda1 + (Q != 1) * lambda2 / (sum_betas(j) + tau2));

                  joint_b_matrix(j, q) = (bj_bar > threshold) * (bj_bar - threshold) +
                                         (bj_bar < -threshold) * (bj_bar + threshold);
                }

                if (summary_betas(j, q) * joint_b_matrix(j, q) < 0)
                  perror("Sign inverse error");
              }
            }
          }

          // Finished initialization
          // Now, iterate over all SNPs in LD
          for (int p = 0; p < num_SNP; p++)
          {
            int j = index_matrix(p, 0);

            double lambda1 = lambda0 + (Ifunc_SNP(j) == 1) * arma::as_scalar(z_matrix.row(j) * temp_lambda_vec);

            for (int q = 0; q < Q; q++)
            {
              if (skip(j, q) == 0)
              {
                if (summary_betas(j, q) != 0.0)
                {
                  double bj_bar = summary_betas(j, q);

                  for (int i = index_matrix(p, 1); i <= index_matrix(p, 2); i++)
                  {
                    bj_bar -= ld_vec(i) * joint_b_matrix(ld_J(i), q);
                  }

                  double threshold = ((z_scale == 1) * SD_vec(j, q) + (z_scale == 0)) *
                                     (lambda1 + (Q != 1) * lambda2 / (sum_betas(j) + tau2));

                  if (fabs(bj_bar) > upper_val)
                  {
                    if (breaking == 1)
                    {
                      num_iter_vec(tuning_index) = -1;
                      break;
                    }
                    else
                    {
                      bj_bar = 0.0;
                      skip(j, q) = 1;
                    }
                  }

                  joint_b_matrix(j, q) = (bj_bar > threshold) * (bj_bar - threshold) +
                                         (bj_bar < -threshold) * (bj_bar + threshold);
                }
                else
                {
                  joint_b_matrix(j, q) = 0.0;
                }
              }
              else
              {
                joint_b_matrix(j, q) = 0.0;
              }
            }
          }

          // Update the skip matrix
          skip.elem(
                  arma::find(
                      arma::abs(joint_b_matrix) > upper_val))
              .ones();

          // Check for divergence
          int df_q = arma::accu(
              arma::abs(joint_b_matrix(arma::span::all, 0)) != 0.0);
          if (df_q > df_max)
            converges = false;

          if (!converges)
          {
            num_iter_vec(tuning_index) = -2;
            break;
          }

          // Check if the beta matrix has converged
          bool found = arma::any(
              arma::find(
                  arma::abs(temp_b_matrix - joint_b_matrix) > epsilon));

          if (!found)
          {
            // Determine which entries need to update
            arma::uvec beta_indices = arma::find(joint_b_matrix != 0.0);
            beta_indices.for_each(
                [&](arma::uword idx)
                {
                  arma::uvec beta_subscript = arma::ind2sub(arma::size(joint_b_matrix), idx);
                  beta_matrix(tuning_index, idx) = joint_b_matrix(beta_subscript(0), beta_subscript(1));
                });

            num_iter_vec(tuning_index) = n;
            break;
          }

          // Update for next iteration if we haven't converged yet
          temp_b_matrix(arma::span::all, arma::span::all) = joint_b_matrix(arma::span::all, arma::span::all);
          sum_betas(arma::span::all) = arma::sum(arma::abs(joint_b_matrix(arma::span::all, arma::span::all)), 1);

          num_iter_vec(tuning_index) = n;
        }
      }
    }
  }
  PutRNGstate();

  return Rcpp::List::create(
      Rcpp::Named("beta_matrix") = beta_matrix,
      Rcpp::Named("num_iter_vec") = num_iter_vec,
      Rcpp::Named("all_tuning_matrix") = all_tuning_matrix);
}
